src/test/java/io/getunleash/impactmetrics/CounterImplTest.java

```diff
@@ -0,0 +1,130 @@
+
+        assertThat(result.getSamples()).hasSize(1);
+        assertThat(result.getSamples().get(0).getValue()).isEqualTo(5L);
+        assertThat(result.getSamples().get(0).getLabels()).containsEntry("foo", "bar");
+    }
+
+    @Test
+    public void should_store_different_label_combinations_separately() {
+        CounterImpl counter = new CounterImpl("multi_label", "label test");
+
+        counter.inc(1, Map.of("a", "x"));
+        counter.inc(2, Map.of("b", "y"));
+        counter.inc(3);
+
+        CollectedMetric result = counter.collect();
+
+        assertThat(result.getSamples()).hasSize(3);
+
+        NumericMetricSample sampleAX =
+                result.getSamples().stream()
+                        .filter(s -> s.getLabels().containsKey("a"))
+                        .findFirst()
+                        .orElseThrow();
+        NumericMetricSample sampleBY =
+                result.getSamples().stream()
+                        .filter(s -> s.getLabels().containsKey("b"))
+                        .findFirst()
+                        .orElseThrow();
+        NumericMetricSample sampleEmpty =
+                result.getSamples().stream()
+                        .filter(s -> s.getLabels().isEmpty())
+                        .findFirst()
+                        .orElseThrow();
+
+        assertThat(sampleAX.getValue()).isEqualTo(1L);
+        assertThat(sampleBY.getValue()).isEqualTo(2L);
+        assertThat(sampleEmpty.getValue()).isEqualTo(3L);
+    }
+
+    @Test
+    public void should_return_zero_value_when_empty() {
+        CounterImpl counter = new CounterImpl("empty_counter", "empty");
+
+        CollectedMetric result = counter.collect();
+
+        assertThat(result.getSamples()).hasSize(1);
+        assertThat(result.getSamples().get(0).getValue()).isEqualTo(0L);
+        assertThat(result.getSamples().get(0).getLabels()).isEmpty();
+    }
+
+    @Test
+    public void should_reset_after_collect() {
+        CounterImpl counter = new CounterImpl("reset_test", "reset");
+
+        counter.inc(5);
+        CollectedMetric first = counter.collect();
+        assertThat(first.getSamples().get(0).getValue()).isEqualTo(5L);
+
+        CollectedMetric second = counter.collect();
+        assertThat(second.getSamples()).hasSize(1);
+        assertThat(second.getSamples().get(0).getValue()).isEqualTo(0L);
+    }
+
+    @Test
+    public void should_generate_stable_label_key() {
+        Map<String, String> labels1 = Map.of("z", "1", "a", "2");
+        Map<String, String> labels2 = Map.of("a", "2", "z", "1");
+
+        String key1 = CounterImpl.getLabelKey(labels1);
+        String key2 = CounterImpl.getLabelKey(labels2);
+
+        assertThat(key1).isEqualTo(key2);
+        assertThat(key1).isEqualTo("a=2,z=1");
+    }
+
+    @Test
+    public void should_parse_label_key_correctly() {
+        String key = "env=prod,region=us-east";
+
+        Map<String, String> labels = CounterImpl.parseLabelKey(key);
```

Comment:
this is private bahavior that should not be exposed and only tested transitively as part of the core behavior